components/net_connect/wifi_connect.c

```diff
@@ -0,0 +1,340 @@
+    ESP_ERROR_CHECK(esp_wifi_clear_default_wifi_driver_and_handlers(s_example_sta_netif));
+    esp_netif_destroy(s_example_sta_netif);
+    s_example_sta_netif = NULL;
+}
+
+
+esp_err_t net_connect_wifi_sta_do_connect(wifi_config_t wifi_config, bool wait)
+{
+    if (wait) {
+        s_semph_get_ip_addrs = xSemaphoreCreateBinary();
+        if (s_semph_get_ip_addrs == NULL) {
+            return ESP_ERR_NO_MEM;
+        }
+#if CONFIG_NET_CONNECT_IPV6
+        s_semph_get_ip6_addrs = xSemaphoreCreateBinary();
+        if (s_semph_get_ip6_addrs == NULL) {
+            vSemaphoreDelete(s_semph_get_ip_addrs);
+            return ESP_ERR_NO_MEM;
+        }
+#endif
+    }
+    s_retry_num = 0;
+    ESP_ERROR_CHECK(esp_event_handler_register(WIFI_EVENT, WIFI_EVENT_STA_DISCONNECTED, &example_handler_on_wifi_disconnect, NULL));
+    ESP_ERROR_CHECK(esp_event_handler_register(IP_EVENT, IP_EVENT_STA_GOT_IP, &example_handler_on_sta_got_ip, NULL));
+    ESP_ERROR_CHECK(esp_event_handler_register(WIFI_EVENT, WIFI_EVENT_STA_CONNECTED, &example_handler_on_wifi_connect, s_example_sta_netif));
+#if CONFIG_NET_CONNECT_IPV6
+    ESP_ERROR_CHECK(esp_event_handler_register(IP_EVENT, IP_EVENT_GOT_IP6, &example_handler_on_sta_got_ipv6, NULL));
+#endif
+
+    ESP_LOGI(TAG, "Connecting to %s...", wifi_config.sta.ssid);
+    ESP_ERROR_CHECK(esp_wifi_set_config(WIFI_IF_STA, &wifi_config));
+    esp_err_t ret = esp_wifi_connect();
+    if (ret != ESP_OK) {
+        ESP_LOGE(TAG, "WiFi connect failed! ret:%x", ret);
+        return ret;
+    }
+    if (wait) {
+        ESP_LOGI(TAG, "Waiting for IP(s)");
+#if CONFIG_NET_CONNECT_IPV4
+        xSemaphoreTake(s_semph_get_ip_addrs, portMAX_DELAY);
+        vSemaphoreDelete(s_semph_get_ip_addrs);
+        s_semph_get_ip_addrs = NULL;
+#endif
+#if CONFIG_NET_CONNECT_IPV6
+        xSemaphoreTake(s_semph_get_ip6_addrs, portMAX_DELAY);
+        vSemaphoreDelete(s_semph_get_ip6_addrs);
+        s_semph_get_ip6_addrs = NULL;
+#endif
+        if (s_retry_num > CONFIG_NET_CONNECT_WIFI_CONN_MAX_RETRY) {
+            return ESP_FAIL;
+        }
+    }
+    return ESP_OK;
+}
+
+esp_err_t net_connect_wifi_sta_do_disconnect(void)
+{
+    ESP_ERROR_CHECK(esp_event_handler_unregister(WIFI_EVENT, WIFI_EVENT_STA_DISCONNECTED, &example_handler_on_wifi_disconnect));
+    ESP_ERROR_CHECK(esp_event_handler_unregister(IP_EVENT, IP_EVENT_STA_GOT_IP, &example_handler_on_sta_got_ip));
+    ESP_ERROR_CHECK(esp_event_handler_unregister(WIFI_EVENT, WIFI_EVENT_STA_CONNECTED, &example_handler_on_wifi_connect));
+#if CONFIG_NET_CONNECT_IPV6
+    ESP_ERROR_CHECK(esp_event_handler_unregister(IP_EVENT, IP_EVENT_GOT_IP6, &example_handler_on_sta_got_ipv6));
+#endif
+    return esp_wifi_disconnect();
+}
+
+void net_connect_wifi_shutdown(void)
+{
+    ESP_LOGI(TAG, "WiFi shutdown handler called");
+    net_connect_wifi_sta_do_disconnect();
+    net_connect_wifi_stop();
+}
+
+net_iface_handle_t net_configure_wifi_sta(net_wifi_sta_config_t *config)
```

Comment:
Now wifi provides ample runtime configuration, while others are more or less Kconfig options. 
should we unify this, or do we prefer runtime to compile-time config?